core/src/main/java/kafka/server/ServerSocketFactory.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.network.Selectable;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.nio.channels.ServerSocketChannel;
+
+public interface ServerSocketFactory {
+    ServerSocketChannel openServerSocket(
+        String listenerName,
+        InetSocketAddress socketAddress,
+        int listenBacklogSize,
+        int recvBufferSize
+    ) throws IOException;
+
+    class KafkaServerSocketFactory implements ServerSocketFactory {
+        public static final KafkaServerSocketFactory INSTANCE = new KafkaServerSocketFactory();
+
+        @Override
+        public ServerSocketChannel openServerSocket(
+                String listenerName,
+                InetSocketAddress socketAddress,
+                int listenBacklogSize,
+                int recvBufferSize
+        ) throws IOException {
+            ServerSocketChannel socketChannel = null;
+            try {
+                socketChannel = ServerSocketChannel.open();
```

Comment:
Could you please rewrite it by following style to avoid resource leak.
```java
            ServerSocketChannel socketChannel = ServerSocketChannel.open();
            try {
                socketChannel.configureBlocking(false);
```

Reply:
Hmm... there shouldn't be a resource leak in the current case where it's initialized to null, since that's handled in the `catch` block. However, I can initialize the socketChannel variable with the result of `ServerSocketChannel.open`. That will make the code shorter and there's no disadvantage that I can see.

Reply:
> there shouldn't be a resource leak in the current case where it's initialized to null, since that's handled in the catch block.

oh, you are right. thanks for sharing.